File: backend/src/main/java/kr/co/softice/mes/domain/service/AlarmService.java
package kr.co.softice.mes.domain.service;

import kr.co.softice.mes.common.exception.BusinessException;
import kr.co.softice.mes.common.exception.ErrorCode;
import kr.co.softice.mes.domain.entity.AlarmHistoryEntity;
import kr.co.softice.mes.domain.entity.AlarmTemplateEntity;
import kr.co.softice.mes.domain.repository.AlarmHistoryRepository;
import kr.co.softice.mes.domain.repository.AlarmTemplateRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Alarm Service
 * 알람 서비스
 *
 * @author Moon Myung-seop
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AlarmService {

    private final AlarmTemplateRepository templateRepository;
    private final AlarmHistoryRepository historyRepository;

    // ==================== Template Management ====================

    /**
     * Find all templates
     */
    @Transactional(readOnly = true)
    public List<AlarmTemplateEntity> findAllTemplates(String tenantId) {
        return templateRepository.findAllByTenantId(tenantId);
    }

    /**
     * Find template by event type
     */
    @Transactional(readOnly = true)
    public Optional<AlarmTemplateEntity> findTemplateByEventType(String tenantId, String eventType) {
        return templateRepository.findByTenantIdAndEventType(tenantId, eventType);
    }

    // ==================== Alarm Sending ====================

    /**
     * Send alarm
     */
    @Transactional
    public AlarmHistoryEntity sendAlarm(
            String tenantId,
            String eventType,
            Long recipientUserId,
            String recipientName,
            Map<String, String> variables,
            String referenceType,
            Long referenceId,
            String referenceNo
    ) {
        log.info("Sending alarm: eventType={}, recipient={}", eventType, recipientUserId);

        // Find template
        AlarmTemplateEntity template = findTemplateByEventType(tenantId, eventType)
                .orElseThrow(() -> {
                    log.warn("No template found for event type: {}", eventType);
                    return new BusinessException(ErrorCode.ALARM_TEMPLATE_NOT_FOUND,
                            "알람 템플릿을 찾을 수 없습니다: eventType=" + eventType);
                });

        // Render message
        String title = template.renderTitle(variables);
        String message = template.renderMessage(variables);

        // Create alarm history
        AlarmHistoryEntity alarm = AlarmHistoryEntity.builder()
                .recipientUserId(recipientUserId)
                .recipientName(recipientName)
                .alarmType(template.getAlarmType())
                .eventType(template.getEventType())
                .priority(template.getPriority())
                .title(title)
                .message(message)
                .referenceType(referenceType)
                .referenceId(referenceId)
                .referenceNo(referenceNo)
                .sentViaEmail(template.isEmailEnabled())
                .sentViaSms(template.isSmsEnabled())
                .sentViaPush(template.isPushEnabled())
                .sentViaSystem(template.isSystemEnabled())
                .sendStatus("PENDING")
                .build();

        alarm.setTenant(template.getTenant());
        alarm.setTemplate(template);

        AlarmHistoryEntity saved = historyRepository.save(alarm);

        // Send via channels (in real implementation, integrate with email/SMS/push services)
        sendViaChannels(saved, template);

        return saved;
    }

    /**
     * Send via channels
     */
    private void sendViaChannels(AlarmHistoryEntity alarm, AlarmTemplateEntity template) {
        try {
            if (template.isEmailEnabled()) {
                sendViaEmail(alarm);
            }

            if (template.isSmsEnabled()) {
                sendViaSms(alarm);
            }

            if (template.isPushEnabled()) {
                sendViaPush(alarm);
            }

            // System notification is always sent
            alarm.markAsSent();
            historyRepository.save(alarm);

        } catch (Exception e) {
            log.error("Failed to send alarm [id={}, type={}, recipient={}]: {}",
                    alarm.getAlarmId(), alarm.getEventType(),
                    alarm.getRecipientUserId(), e.getMessage(), e);
            alarm.markAsFailed(e.getMessage());
            historyRepository.save(alarm);
        }
    }

    /**
     * Send alarm via email channel.
     * When an external email service (e.g., SMTP, SES) is configured,
     * replace the placeholder logic below with the actual integration call.
     */
    private void sendViaEmail(AlarmHistoryEntity alarm) {
        log.info("Sending email notification [alarmId={}, recipient={}, email={}, title={}]",
                alarm.getAlarmId(), alarm.getRecipientName(),
                alarm.getRecipientEmail(), alarm.getTitle());
        alarm.setSentViaEmail(true);
    }

    /**
     * Send alarm via SMS channel.
     * When an external SMS gateway is configured,
     * replace the placeholder logic below with the actual integration call.
     */
    private void sendViaSms(AlarmHistoryEntity alarm) {
        log.info("Sending SMS notification [alarmId={}, recipient={}, phone={}, title={}]",
                alarm.getAlarmId(), alarm.getRecipientName(),
                alarm.getRecipientPhone(), alarm.getTitle());
        alarm.setSentViaSms(true);
    }

    /**
     * Send alarm via push notification channel.
     * When an external push service (e.g., FCM, APNs) is configured,
     * replace the placeholder logic below with the actual integration call.
     */
    private void sendViaPush(AlarmHistoryEntity alarm) {
        log.info("Sending push notification [alarmId={}, recipientUserId={}, title={}]",
                alarm.getAlarmId(), alarm.getRecipientUserId(), alarm.getTitle());
        alarm.setSentViaPush(true);
    }

    // ==================== Alarm History Management ====================

    /**
     * Find alarms by recipient
     */
    @Transactional(readOnly = true)
    public List<AlarmHistoryEntity> findAlarmsByRecipient(String tenantId, Long userId) {
        return historyRepository.findByRecipient(tenantId, userId);
    }

    /**
     * Find unread alarms by recipient
     */
    @Transactional(readOnly = true)
    public List<AlarmHistoryEntity> findUnreadAlarms(String tenantId, Long userId) {
        return historyRepository.findUnreadByRecipient(tenantId, userId);
    }

    /**
     * Find recent alarms (last 7 days)
     */
    @Transactional(readOnly = true)
    public List<AlarmHistoryEntity> findRecentAlarms(String tenantId, Long userId) {
        LocalDateTime since = LocalDateTime.now().minusDays(7);
        return historyRepository.findRecentByRecipient(tenantId, userId, since);
    }

    /**
     * Count unread alarms
     */
    @Transactional(readOnly = true)
    public Long countUnreadAlarms(String tenantId, Long userId) {
        return historyRepository.countUnreadByRecipient(tenantId, userId);
    }

    /**
     * Mark alarm as read
     */
    @Transactional
    public void markAsRead(Long alarmId) {
        Optional<AlarmHistoryEntity> alarmOpt = historyRepository.findById(alarmId);
        if (alarmOpt.isPresent()) {
            AlarmHistoryEntity alarm = alarmOpt.get();
            alarm.markAsRead();
            historyRepository.save(alarm);
        }
    }

    /**
     * Mark all alarms as read for user
     */
    @Transactional
    public void markAllAsRead(String tenantId, Long userId) {
        List<AlarmHistoryEntity> unreadAlarms = findUnreadAlarms(tenantId, userId);
        for (AlarmHistoryEntity alarm : unreadAlarms) {
            alarm.markAsRead();
            historyRepository.save(alarm);
        }
    }

    /**
     * Delete old alarms (retention policy)
     */
    @Transactional
    public void deleteOldAlarms(String tenantId, int retentionDays) {
        LocalDateTime before = LocalDateTime.now().minusDays(retentionDays);
        historyRepository.deleteOldAlarms(tenantId, before);
        log.info("Deleted alarms older than {} days for tenant: {}", retentionDays, tenantId);
    }

    // ==================== Statistics ====================

    /**
     * Get alarm statistics
     */
    @Transactional(readOnly = true)
    public AlarmStatistics getStatistics(String tenantId, Long userId) {
        Long unread = countUnreadAlarms(tenantId, userId);
        Long total = (long) findAlarmsByRecipient(tenantId, userId).size();

        Long approvals = historyRepository.countByTypeAndRecipient(tenantId, userId, "APPROVAL");
        Long quality = historyRepository.countByTypeAndRecipient(tenantId, userId, "QUALITY");
        Long production = historyRepository.countByTypeAndRecipient(tenantId, userId, "PRODUCTION");
        Long inventory = historyRepository.countByTypeAndRecipient(tenantId, userId, "INVENTORY");

        return new AlarmStatistics(unread, total, approvals, quality, production, inventory);
    }

    /**
     * Alarm statistics inner class
     */
    @lombok.Data
    @lombok.AllArgsConstructor
    public static class AlarmStatistics {
        private Long unreadCount;
        private Long totalCount;
        private Long approvalCount;
        private Long qualityCount;
        private Long productionCount;
        private Long inventoryCount;

        public Long getReadCount() {
            return totalCount - unreadCount;
        }

        public double getReadRate() {
            if (totalCount == 0) return 0.0;
            return (double) getReadCount() / totalCount * 100;
        }
    }
}

File: frontend/src/services/alarmService.ts
/**
 * Alarm Service
 * 알람 관리 API 서비스
 */

import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8080/api';

// ==================== Interfaces ====================

export interface AlarmTemplate {
  templateId: number;
  tenantId: string;
  templateCode: string;
  templateName: string;
  alarmType: string;
  eventType: string;
  titleTemplate: string;
  messageTemplate: string;
  enableEmail: boolean;
  enableSms: boolean;
  enablePush: boolean;
  enableSystem: boolean;
  priority: 'LOW' | 'NORMAL' | 'HIGH' | 'URGENT';
  isActive: boolean;
}

export interface AlarmHistory {
  alarmId: number;
  tenantId: string;
  recipientUserId: number;
  recipientName?: string;
  alarmType: string;
  eventType: string;
  priority: string;
  title: string;
  message: string;
  referenceType?: string;
  referenceId?: number;
  referenceNo?: string;
  sentViaEmail: boolean;
  sentViaSms: boolean;
  sentViaPush: boolean;
  sentViaSystem: boolean;
  isRead: boolean;
  readAt?: string;
  sendStatus: 'PENDING' | 'SENT' | 'FAILED';
  sentAt?: string;
  createdAt: string;
}

export interface AlarmStatistics {
  unreadCount: number;
  totalCount: number;
  approvalCount: number;
  qualityCount: number;
  productionCount: number;
  inventoryCount: number;
  readCount: number;
  readRate: number;
}

// ==================== Alarm Service ====================

class AlarmServiceClass {
  private baseURL = `${API_BASE_URL}/alarms`;

  /**
   * Get all templates
   */
  async getAllTemplates(tenantId: string): Promise<AlarmTemplate[]> {
    const response = await axios.get(`${this.baseURL}/templates`, {
      params: { tenantId }
    });
    return response.data.data;
  }

  /**
   * Get alarms for user
   */
  async getAlarms(tenantId: string, userId: number): Promise<AlarmHistory[]> {
    const response = await axios.get(this.baseURL, {
      params: { tenantId, userId }
    });
    return response.data.data;
  }

  /**
   * Get unread alarms
   */
  async getUnreadAlarms(tenantId: string, userId: number): Promise<AlarmHistory[]> {
    const response = await axios.get(`${this.baseURL}/unread`, {
      params: { tenantId, userId }
    });
    return response.data.data;
  }

  /**
   * Get recent alarms (last 7 days)
   */
  async getRecentAlarms(tenantId: string, userId: number): Promise<AlarmHistory[]> {
    const response = await axios.get(`${this.baseURL}/recent`, {
      params: { tenantId, userId }
    });
    return response.data.data;
  }

  /**
   * Count unread alarms
   */
  async countUnreadAlarms(tenantId: string, userId: number): Promise<number> {
    const response = await axios.get(`${this.baseURL}/unread/count`, {
      params: { tenantId, userId }
    });
    return response.data.data;
  }

  /**
   * Mark alarm as read
   */
  async markAsRead(alarmId: number): Promise<void> {
    await axios.put(`${this.baseURL}/${alarmId}/read`);
  }

  /**
   * Mark all alarms as read
   */
  async markAllAsRead(tenantId: string, userId: number): Promise<void> {
    await axios.put(`${this.baseURL}/read-all`, null, {
      params: { tenantId, userId }
    });
  }

  /**
   * Get alarm statistics
   */
  async getStatistics(tenantId: string, userId: number): Promise<AlarmStatistics> {
    const response = await axios.get(`${this.baseURL}/statistics`, {
      params: { tenantId, userId }
    });
    return response.data.data;
  }
}

// ==================== Helper Functions ====================

/**
 * Get alarm type label
 */
export function getAlarmTypeLabel(type: string): string {
  const labels: Record<string, string> = {
    'SYSTEM': '시스템',
    'APPROVAL': '결재',
    'QUALITY': '품질',
    'PRODUCTION': '생산',
    'INVENTORY': '재고',
    'DELIVERY': '출하'
  };
  return labels[type] || type;
}

/**
 * Get alarm type color
 */
export function getAlarmTypeColor(type: string): 'default' | 'primary' | 'success' | 'warning' | 'error' {
  const colors: Record<string, 'default' | 'primary' | 'success' | 'warning' | 'error'> = {
    'SYSTEM': 'default',
    'APPROVAL': 'primary',
    'QUALITY': 'warning',
    'PRODUCTION': 'success',
    'INVENTORY': 'error',
    'DELIVERY': 'primary'
  };
  return colors[type] || 'default';
}

/**
 * Get priority label
 */
export function getPriorityLabel(priority: string): string {
  const labels: Record<string, string> = {
    'LOW': '낮음',
    'NORMAL': '보통',
    'HIGH': '높음',
    'URGENT': '긴급'
  };
  return labels[priority] || priority;
}

/**
 * Get priority color
 */
export function getPriorityColor(priority: string): 'default' | 'primary' | 'warning' | 'error' {
  const colors: Record<string, 'default' | 'primary' | 'warning' | 'error'> = {
    'LOW': 'default',
    'NORMAL': 'primary',
    'HIGH': 'warning',
    'URGENT': 'error'
  };
  return colors[priority] || 'default';
}

/**
 * Format date time
 */
export function formatDateTime(dateStr?: string): string {
  if (!dateStr) return '-';
  const date = new Date(dateStr);
  return date.toLocaleString('ko-KR', {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit'
  });
}

/**
 * Get relative time
 */
export function getRelativeTime(dateStr: string): string {
  const date = new Date(dateStr);
  const now = new Date();
  const diff = now.getTime() - date.getTime();

  const minutes = Math.floor(diff / 60000);
  const hours = Math.floor(diff / 3600000);
  const days = Math.floor(diff / 86400000);

  if (minutes < 1) return '방금 전';
  if (minutes < 60) return `${minutes}분 전`;
  if (hours < 24) return `${hours}시간 전`;
  if (days < 7) return `${days}일 전`;
  return formatDateTime(dateStr);
}

// Export service instance
export const alarmService = new AlarmServiceClass();
